### PR TITLE
[DataFrame] Handling zero length partitions

### PR DESCRIPTION
**Pull request by kunalgosar**
This filters out zero length partitions on getting `col_partitions` or `row_partitions`.

### CODE REVIEW

File: python/ray/dataframe/groupby.py [hunk 1]
```diff
@@ -208,6 +209,9 @@ def apply_helper(df):
                     num_return_vals=len(new_df._block_partitions))
                     for block in new_df._block_partitions.T]).T
                 new_df.index = self._index
+                new_df._row_metadata = \
+                    _IndexMetadata(new_df._block_partitions[:, 0],
```
Comment (pschafhalter) [inline python/ray/dataframe/groupby.py]: The `concat` from earlier constructs a new `DataFrame` with a new `_IndexMetadata` which gets overwritten here. Ideally, we could modify concat to pass this `_indexMetadata` to the  `DataFrame` constructor, but I'm not sure if this is possible to do elegantly with the current code base.

No need to make a change to the PR, but would like to point this out.

**Review by pschafhalter — CHANGES_REQUESTED** [review]

File: python/ray/dataframe/groupby.py (hunk 1 above)
Comment (kunalgosar) [reply python/ray/dataframe/groupby.py]: The issue here is that the reindex that is applied on columns changes the partition structure and the old `_IndexMetadata` created on `concat` is invalid. I agree that duplicate work is being done here, but I'm not sure there is a nice solution.

**Review by pschafhalter — APPROVED** [review]

**Comment by kunalgosar** [comment]
Jenkins, retest this please

**Comment by kunalgosar** [comment]
Jenkins, retest this please

**Comment by kunalgosar** [comment]
Jenkins, retest this please

**Comment by kunalgosar** [comment]
Jenkins, retest this please

**Comment by kunalgosar** [comment]
Jenkins, retest this please

**Comment by devin-petersohn** [comment]
Jenkins, retest this please.

**Review by devin-petersohn — CHANGES_REQUESTED** [review]
Left a few comment. Thanks for the fix!

**Comment by kunalgosar** [comment]
Jenkins, retest this please.

**Comment by devin-petersohn** [comment]
Jenkins, retest this please.

**Review by devin-petersohn — APPROVED** [review]

**Pull request merged**